### PR TITLE
feat(rtc): v1 self-hosted real-time collaboration

### PR DESCRIPTION
**Pull request by neotyk**
See [Project Overview](https://github.com/athensresearch/athens/projects/16)

Merging #1117 into this, see https://github.com/neotyk/athens/pull/2

### CODE REVIEW

File: src/cljc/athens/common_events.cljc [hunk 1]
```diff
@@ -0,0 +1,17 @@
+(ns athens.common-events
+  "Event as Verbs executed on Knowledge Graph"
+  (:require [clojure.string :as string]))
+
+(defn paste-verbatim->tx [uid text start value]
```
Comment (agentydragon) [inline src/cljc/athens/common_events.cljc:5]: Consider adding a test.

**Comment by neotyk** [comment]
Proposed event flow

> - We 1st receive `re-frame` event (that already is a verb)
> - if remote
>     - create `event-id`
>     - read `last-seen-tx`
>     - start waiting for `event-id` ack
>         - ui: event queued
>     - send via #[[WebSocket]] event
>         - with `event-id, last-seen-tx, event-args`
>     - on server: receive event
>         - check `last-seen-tx`
>             - reject event when potentially conflicting
>                 - let client know what is the `last-global-tx`
>             - working from current head
>                 - resolve event in db context into txs (**common-events**)
>                 - `transact!`
>                 - send via #[[WebSocket]] event ack
>                 - broadcast tx log to connected clients
>     - on client
>         - when ack received
>             - remove `event-id` from awaiting ack
>             - ui: event completed
>         - when rejection received
>             - ui: notify user of unsuccessful event execution
>             - update `last-global-tx`
>             - ui: lagging db
>                 - subscription `::subs/remote-ahead?`
>                     - `last-seen-tx < last-global-tx`
> - else
>     - resolve event in local db context into txs (**common-events**)
>     - `transact!`
> - on client: when received `tx-log` broadcast
>     - apply received `tx-log` to local `DataScript` db
>     - ui: automatic update via reactive
>

**Comment by sid597** [comment]
I have a few questions regarding this :
1. How does the server know some device has disconnected? 
2. Can a client have stale data because of a network partition between the server and that client? I think the answer to this question depends on how question 1 is handled.

**Comment by tangjeff0** [comment]
> How does the server know some device has disconnected?

Websocket servers typically provide a very easy API for detecting `connect` and `disconnect` events. @sid597

**Comment by sid597** [comment]
Thank you @tangjeff0

I will have to read more regarding the Websocket approach, questions I will have in mind while reading it are : 
- Is network partition the same as `disconnect`? (assuming network will heal )
- Difference between network delay and disconnect, how is it determined what case is what.
- In case of network delays :
    - How is the situation of transactions arriving out of order is addresed?
    - Does the library resend a transaction, if so in what case and how would same transaction arriving on the client-side handled?

**Comment by tangjeff0** [comment]
> Is network partition the same as disconnect? (assuming network will heal )

Yes. Heal = reconnect

> Difference between network delay and disconnect, how is it determined what case is what.

If there is delay but no disconnect, should be handled by TCP or networking layer.

[Offline is a special case of lag compensation where lag is infinite. Offline mode should look just like a normal app without real-time updates. ](https://tonsky.me/blog/the-web-after-tomorrow/#:~:text=offline%20is%20a%20special%20case%20of%20lag%20compensation%20where%20lag%20is%20infinite.%20offline%20mode%20should%20look%20just%20like%20a%20normal%20app%20without%20real-time%20updates.%20)

> How is the situation of transactions arriving out of order is addresed?

Transactions should arrive in order, should be handled by TCP layer or websocket library.

**Comment by sid597** [comment]
Now that all the baseless assumptions and worries I had are resolved, how can I help in this implementation @neotyk @tangjeff0

File: src/cljs/athens/events.cljs [hunk 2]
```diff
@@ -1698,28 +1699,34 @@
...
-                                      text
-                                      (subs value start)))
-         tx-data [{:db/id        [:block/uid uid]
-                   :block/string new-string}]]
-     {:dispatch [:transact tx-data]})))
+         local?                (not (:db/remote-graph-conf db))]
+     (if local?
+       {:dispatch [:transact (common-events/paste-verbatim->tx uid text start value)]}
+       ;; TODO event handler for `:remote/paste-verbatim`
+       {:dispatch [:remote/paste-verbatim uid text start value]}))))
```
Comment (tangjeff0) [inline src/cljs/athens/events.cljs:1729]: So every datascript+re-frame transaction needs to check if `local?`?

Is this a good use case for [interceptors](https://day8.github.io/re-frame/Interceptors/)? Every `ds+rf` event can have a `local` interceptor that decides to write transaction on client datascript or through backend datahike.

File: src/clj/athens/self_hosted/web/presence.clj [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+(defn presence-handler
+  [req]
+  (http/as-channel req
+                   {:on-open    #'open-handler
+                    :on-receive #'receive-handler
+                    :on-close   #'close-handler}))
+
+(compojure/defroutes presence-routes
+  (compojure/GET "/ws-presence" [] presence-handler))
```
Comment (tangjeff0) [inline src/clj/athens/self_hosted/web/presence.clj:79]: Please add `(compojure/GET "/ws-presence" [] <h1>hello world</h1>)` for easy check that server is up.

**Review by tangjeff0 — COMMENTED** [review]
Awesome job so far @neotyk. See comments about interceptor

File: dev/clj/user.clj [hunk 4]
```diff
@@ -0,0 +1,25 @@
+(ns user
+  (:require [athens.self-hosted.core      :as app]
+            [clojure.tools.namespace.repl :as repl]
+            [com.stuartsierra.component   :as component]))
```
[the remark this answers is not in the document]
Comment (neotyk) [reply dev/clj/user.clj:4]: And component works on both clj & cljs.

File: src/cljs/athens/events.cljs (hunk 2 above)
Comment (neotyk) [reply src/cljs/athens/events.cljs:1729]: I need to check, haven't had the need for using interceptors in past.

Comment (neotyk) [reply src/cljs/athens/events.cljs:1729]: Ok, checked.
Using interceptors to "route" events would be possible (through self modifying property of interceptors)
but it would be confusing, because we'd register event-handler `:paster-verbatim` with this magic interceptor,
and magic interceptor would re-route event to other handler, this would result in very unintuitive code.

We could inject cofx to test if we're in local vs self-hosted mode.

Comment (tangjeff0) [reply src/cljs/athens/events.cljs:1732]: Cofx is probably the right move then! Both Cofx and Interceptors are good at reusing some code for events, but Interceptors are a little too magical.

File: src/cljs/athens/events.cljs [hunk 5]
```diff
@@ -1818,3 +1824,93 @@
...
+       :fx [[:dispatch-n events]]})))
+
+
+(reg-event-db
+  :remote/reject-event
+  (fn [db [_ {:keys [event-id reason data] :as rejection-event}]]
+    (js/console.log "reject event" (pr-str rejection-event))
+    (let [awaited-event  (->> (:remote/awaited-events db)
+                              (filter #(= event-id (:event/id %)))
+                              first)
```
Comment (agentydragon) [inline src/cljs/athens/events.cljs:1879]: duplicated with lines 1854-1856, consider deduplicating

(e.g.: `(first-awaited-event db)`)

Comment (agentydragon) [reply src/cljs/athens/events.cljs:1879]: also lines 1893..1895

File: src/cljs/athens/self_hosted/client.cljs [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+    (js/console.log "User online:" username)))
+
+
+(defn- server-event-handler
+  [{:event/keys [id last-tx type args] :as packet}]
+  (js/console.log "<-" id ", last-tx:" last-tx ", type:" type)
+  (js/console.debug "WSClient: server event:" (pr-str packet))
+  (if (schema/valid-server-event? packet)
+
+    (condp = type
```
Comment (agentydragon) [inline src/cljs/athens/self_hosted/client.cljs:241]: `case`? https://clojuredocs.org/clojure.core/case

File: test/athens/events/awaited_events_test.cljs [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    (let [awaited-events (rf/subscribe [:remote/awaited-events])]
+      (is (not (nil? @awaited-events)))
+      (is (empty? @awaited-events)))))
+
+
+(deftest awaiting-one-event
+  (rf-test/run-test-async
+    (fixture/test-fixtures)
+    (rf/dispatch-sync [:boot/web])
+    (let [awaited-events (rf/subscribe [:remote/awaited-events])]
```
Comment (agentydragon) [inline test/athens/events/awaited_events_test.cljs:29]: might be useful to share this setup boilerplate to make each test a bit shorter.

Comment (agentydragon) [reply test/athens/events/awaited_events_test.cljs:29]: also might be useful to factor into one function the test of "is this an empty collection" (e.g., like on lines 32,33), or the test of "is a collection containing this single element"

File: src/clj/athens/self_hosted/components/web.clj [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+
+
+(defn make-handler
+  [datahike]
+  (compojure/routes health-check-route
+                    (make-ws-route datahike)))
+
+
+(defrecord WebServer
+  [config httpkit datahike]
```
Comment (tangjeff0) [inline src/clj/athens/self_hosted/components/web.clj:91]: Why is this ordered `[config httpkit datahike]`? I would expect this to be ordered `[config datahike httpkit]`? because config and datahike are dependencies and httpkit is added in constructor.

File: src/clj/athens/self_hosted/components/web.clj [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+                                            result)))))))))
+
+
+(defn- make-websocket-handler
+  [datahike]
+  (fn websocket-handler
+    [request]
+    (http/as-channel request
+                     {:on-open    #'open-handler
+                      :on-close   #'close-handler
```
Comment (tangjeff0) [inline src/clj/athens/self_hosted/components/web.clj:69]: Why pass symbol instead of function?

File: src/cljs/athens/events.cljs [hunk 10]
```diff
@@ -1,48 +1,79 @@
...
      :local-storage/set! ["db/filepath" filepath]}))
 
 
-(reg-event-db
+(rf/reg-event-fx
+  :remote/connect!
+  (fn [{:keys [db]} [_ connection-config]]
+    (js/console.log ":remote/connect!" (pr-str connection-config))
+    {:db                     (-> db
+                                 (dissoc :db/filepath)
```
Comment (tangjeff0) [inline src/cljs/athens/events.cljs:47]: TODO: unlisten `:db/filepath` node.js file listener

File: src/cljs/athens/effects.cljs [hunk 11]
```diff
@@ -369,21 +374,64 @@
...
+(rf/reg-fx
   :right-sidebar/scroll-top
   (fn []
     (let [right-sidebar (js/document.querySelector ".right-sidebar-content")]
       (when right-sidebar
         (set! (.. right-sidebar -scrollTop) 0)))))
+
+
+;; TODO: temporary, and limits to one client opened.
+(def self-hosted-client (atom nil))
```
Comment (tangjeff0) [inline src/cljs/athens/effects.cljs:398]: I know this is temporary, but is this the same as `ws-connection`?

https://github.com/athensresearch/athens/blob/63aa3515ad71012f32ea0ff1776093e7c6cc1a62/src/cljs/athens/self_hosted/client.cljs#L18

File: src/cljs/athens/self_hosted/client.cljs [hunk 12]
```diff
@@ -0,0 +1,449 @@
...
+    (.removeEventListener "open" open-handler)))
+
+
+(defn- close-handler
+  [event]
+  (js/console.log "WSClient Disconnected:" event)
+  (let [connection (.-target event)
+        url        (.-url connection)]
+    (remove-listeners! connection)
+    (delayed-reconnect! url)))
```
Comment (tangjeff0) [inline src/cljs/athens/self_hosted/client.cljs:340]: Why reconnect on close?

File: src/cljs/athens/effects.cljs [hunk 13]
```diff
@@ -1,23 +1,28 @@
...
     [day8.re-frame.async-flow-fx]
-    [goog.dom.selection :refer [setCursorPosition]]
-    [posh.reagent :as p :refer [transact!]]
-    [re-frame.core :refer [dispatch reg-fx subscribe]]
-    [stylefy.core :as stylefy]))
+    [goog.dom.selection          :refer [setCursorPosition]]
+    [malli.core                  :as m]
+    [malli.error                 :as me]
+    [posh.reagent                :as p :refer [transact!]]
+    [re-frame.core               :as rf]
```
Comment (tangjeff0) [inline src/cljs/athens/effects.cljs:24]: I think we should refactor `[re-frame.core :refer [dispatch reg-fx subscribe]]` to `[re-frame.core :as rf]` in a separate PR here and in `events.cljs`.

File: src/clj/athens/self_hosted/components/web.clj (hunk 9 above)
Comment (tangjeff0) [reply src/clj/athens/self_hosted/components/web.clj:69]: `#'open-handler` - a ref, can develop and live reload this value. run-time. refer to current value

`open-handler` - refer to compile-time function. always refer to old one

File: src/clj/athens/self_hosted/components/web.clj (hunk 8 above)
Comment (tangjeff0) [reply src/clj/athens/self_hosted/components/web.clj:91]: `defrecord` semantics destructure the map based on key name.

Comment (tangjeff0) [reply src/clj/athens/self_hosted/components/web.clj:91]: Order doesn't matter, but it does make sense to order meta-data first (I'm right :))

File: src/cljs/athens/effects.cljs (hunk 11 above)
Comment (tangjeff0) [reply src/cljs/athens/effects.cljs:398]: `component` and `re-frame` both frameworks. This is first instance of using `component` on front-end. How to use together if at all?

File: src/cljs/athens/self_hosted/client.cljs (hunk 12 above)
Comment (tangjeff0) [reply src/cljs/athens/self_hosted/client.cljs:340]: `on-close` means we didn't try to stop the connection, it was unintentional like a network issue or closing laptop, so we should try to reconnect.

There is a specific close event that should only be called intentionally.

**Comment by tangjeff0** [comment]
Some notes and TODOs from https://github.com/neotyk/athens/pull/2:

- [ ] generate colors and names when a user joins
- [ ] `:presence/all-online` is sent a user joins, but should probably also send `:presence/all-editing`, otherwise need to wait until someone clicks for the first presence uid to be shared
- [ ] delete old presence code
- [ ] convert `:presence/users` from vector into map


Later
- [ ] convert namespaced keys into nested map, e.g. `:presence/users` into `:presence {:users ...}`

File: src/cljs/athens/effects.cljs [hunk 14]
```diff
@@ -270,6 +271,10 @@
 (rf/reg-fx
   :transact!
   (fn [tx-data]
+    ;; 🎶 Sia "Cheap Thrills"
```
Comment (sid597) [inline src/cljs/athens/effects.cljs:274]: 😆

File: src/cljs/athens/events/remote.cljs [hunk 15]
```diff
@@ -0,0 +1,458 @@
...
+
+(rf/reg-event-fx
+  :remote/connect!
+  (fn [{:keys [db]} [_ connection-config]]
+    (js/console.log ":remote/connect!" (pr-str connection-config))
+    {:db                     (-> db
+                                 (dissoc :db/filepath)
+                                 (assoc :db/remote connection-config))
+     :remote/client-connect! connection-config
+     :local-storage/set!     ["db/remote" connection-config]
```
Comment (sid597) [inline src/cljs/athens/events/remote.cljs:22]: @neotyk Connection config is a map so why are we not serializing it while saving in local storage?

Comment (neotyk) [reply src/cljs/athens/events/remote.cljs:22]: Because didn't implement config, yet
